docs(rfc): add vex support (#326)
rancher-sandbox/sbombastic: docs/rfc/0003_vex_support.md

This registry will host a specific kind of images depending on the scope of the 
department. In order to make the scan as accurate as possible, we are going to 
provide a new clusterwide CRD called VexHub. This CRD will hold the 
configuration of the VexHub repositories, so that the Registry CRD will refer 
to these VexHub resources.

Since VexHub CRD is clusterwide, this means that you can use the same 
configuration across multiple registries.

## VexHub CRD

To configure a new VexHub, the user will need to apply a manifest as follows:

```yaml
apiVersion: sbombastic.rancher.io/v1alpha1
kind: VexHub
metadata:
  name: vendor-vexhub
  namespace: default
spec:
  url: https://vexhub.vendor.com
  enabled: true
  credentials:
    secretRef:
      name: vendor-vexhub-secret
      key: access-token
```

There are few attributes to keep in mind when configuring the `VexHub` CRD:

* `url`: is the url of the VexHub repository (this is mandatory)

* `enabled`: to enable/disable the VexHub repo (`true` by default)

* `credentials`: used to store auth credentials to get access to the repo

As you can see, the VexHub CRD has a `secretRef` field, which means that you 

> need to provide a Secret to configure the credentials (if any):

Author: I've improved the description, this was a bit confused.

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: vexhub-auth
  namespace: security-team
type: Opaque
data:
  username: <username>
  token: <base64-encoded-token>
```

Or using `username` and `password` as in this example:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: vexhub-auth
  namespace: security-team
type: Opaque
data:
  username: <username>
  password: <password>
```

This way, the registry will be scanned with VEX files.

In addition to this, we have to consider the fact that VexHub repositories
can use HTTPS encryption. This means that user might want to use their own 

> certificates to verify the connection of the repository.

Author: I was vague there because we are not sure we can set this parameters on `trivy` at the moment.
As far as we know we can set this as environment variable, so this will involve the whole **worker**.
For that reason we were thinking of giving the possibility to the user to configure it at worker level through the helm chart.
cc @fabriziosestito (p.s. can you please link the trivy issue here, I'm not able to find it)

> certificates to verify the connection of the repository.

Reviewer: @alegrey91, the links are in this comment already:  https://github.com/rancher-sandbox/sbombastic/pull/326/files#r2215691918

Additionally, Trivy has a `TRIVY_INSECURE` environment variable, but it's only used for registry connections https://github.com/aquasecurity/trivy/blob/1163b044c7e91a81bba3a862cc4a38e90182f0b4/docs/docs/references/troubleshooting.md?plain=1#L75. 

We need to investigate this further. I am ok moving this to the unknowns section.

> certificates to verify the connection of the repository.

Author: moved under Unresolved questions. Thanks for the suggestion.

The use of this format can sensitively reduce the number of vulnerabilities 
in the final vulnerability report, which is more often full of false positives
entries.

To reduce the noise, VEX is the right choice if you are dealing with SBOMs and 
OCI images. Vex Hub repositories are a collection of VEX files with a defined
structure so that scanning tools can pull them locally and use them to filter
out their output.

That said, we want to add support for:

* Keeping the configuration of multiple Vex Hub repositories

* Support private VexHub repos with credentials and auth tokens

* Use VEX documents to scan images

## Examples / User Stories

[examples]: #examples

<!---
Examples of how the feature will be used. Interactions should show the action
and the response. When appropriate, provide user stories in the form of "As a
[role], I want [feature], so [that]."
--->

### User story #1

As a user, I want to scan registries on my infrastructure filtering out the 
number of CVEs detected by SBOMbastic, removing false positives as much as possible.

### User story #2

As a user, I want to configure the registry scan with appropriate VEX files 

> depending on the content of the registry (eg. test, staging, prod images), 

Reviewer: ```suggestion
depending on the content of the registry (e.g., test, staging, prod images), 
```

so that the vulnerability report will have an accurate result.

# Detailed design

[design]: #detailed-design

<!---
This is the bulk of the RFC. Explain the design in enough detail for somebody
familiar with the product to understand, and for somebody familiar with the
internals to implement.

This section should cover architecture aspects and the rationale behind
disruptive technical decisions (when applicable), as well as corner-cases and
warnings.
--->

This RFC introduces support for VEX files during registry scanning.
From now on, the Registry CRD will have an additional field to refer to the 

> `VexHub` resources we want to use to scan it.

Reviewer: we could call the CRD `VEXHub` instead

Suppose you are scanning the registry of a dedicated department in your company.
This registry will host a specific kind of images depending on the scope of the 
department. In order to make the scan as accurate as possible, we are going to 
provide a new clusterwide CRD called `VexHub`. This CRD will hold the 
configuration of the Vex Hub repositories, so that the Registry CRD will refer 
to these `VexHub` resources.

Since VexHub CRD is clusterwide, this means that you can use the same 
configuration across multiple registries.

## VexHub CRD

To configure a new `VexHub` resource, the user will need to apply a manifest 
as follows:

```yaml
apiVersion: sbombastic.rancher.io/v1alpha1
kind: VexHub
metadata:
  name: company-vexhub
  namespace: default
spec:
  url: "https://vex.company.com/"
  enabled: true
  authSecret:
    name: company-vexhub-creds
    namespace: default
```

There are few attributes to keep in mind when configuring the `VexHub` CRD:

> * `url`: is the url of the Vex Hub repository (this is mandatory)

Reviewer: ```suggestion
* `url`: is the URL of the Vex Hub repository (this is mandatory)
```

* `enabled`: to enable/disable the Vex Hub repo (`true` by default)

* `authSecret`: used to store auth secret to get access to the repo (optional)

A VEX Hub repository might have authentication enforced. When that happens, 

> the `VexHub` CRD has a `spec.credentials.secretRef` field which is used to 

Reviewer: ```suggestion
the `VexHub` CRD has a `spec.credentials.secretRef` field, which is used to 
```

> reference a `Secret` resource that holds the authentication details.

Reviewer 2: this has to be updated with the new `authSecret` field you introduced

```yaml
# Example Secret for Basic Authentication
apiVersion: v1
kind: Secret
metadata:
  name: company-vexhub-creds
  namespace: default
type: Opaque
data:
  username: YWRtaW4=
  password: c2VjcmV0cGFzc3dvcmQ=
```

Or using `username` and `password` as in this example:

```yaml
# Example Secret for Bearer Token
apiVersion: v1
kind: Secret
metadata:
  name: company-vexhub-token
  namespace: default
type: Opaque
data:
  token: ZXlKaGaU9pSktWMVFpT2c9PQ==
```

> The user can use these two fomats for secrets, since Vex Hub supports auth

Reviewer: nit(here and in the rest of the document): it's VEX not Vex

> The user can use these two fomats for secrets, since Vex Hub supports auth

Reviewer: ```suggestion
The user can use these two formats for secrets, since Vex Hub supports auth
```

> throught `username`/`password` or `token`.

Reviewer: ```suggestion
through `username`/`password` or `token`.
```

> To ensure the validity of the input, we are going to provide a webhook to 

Reviewer: we can be more specific (e.g. a `ValidatingWebhook`)

ensure that the referenced `Secret` has the right form.

In addition to this, we have to consider the fact that Vex Hub repositories
can use HTTPS encryption. This means that user might want to use their own 
certificates to verify the connection of the repository.

## AirGap

AirGap is available by default for this feature, since the only requirement is
to provide a self-hosted vexhub repository and change the `repository_url` 

> (if any) whitin the VEX files, to point to the internal registries.

Reviewer: ```suggestion
(if any) within the VEX files, to point to the internal registries.
```

department. In order to make the scan as accurate as possible, we are going to 
provide a new clusterwide CRD called `VEXHub`. This CRD will hold the 
configuration of the VEX Hub repositories, so that the Registry CRD will refer 
to these `VEXHub` resources.

Since VEXHub CRD is clusterwide, this means that you can use the same 
configuration across multiple registries.

## VEXHub CRD

To configure a new `VEXHub` resource, the user will need to apply a manifest 
as follows:

```yaml
apiVersion: sbombastic.rancher.io/v1alpha1
kind: VEXHub
metadata:
  name: company-vexhub
  namespace: default
spec:
  url: "https://vex.company.com/"
  enabled: true
  authSecret:
    name: company-vexhub-creds
    namespace: default
```

There are a few attributes to keep in mind when configuring the `VEXHub` CRD:

* `url`: is the URL of the VEX Hub repository (this is mandatory)

* `enabled`: to enable/disable the VEX Hub repo (`true` by default)

* `authSecret`: used to store auth secret to get access to the repo (optional)

A VEX Hub repository might have authentication enforced. When that happens, 
the `VEXHub` CRD has a `spec.authSecret` field, which is used to 

> reference a `Secret` resource that holds the authentication details.

Reviewer 2: Maybe it's worth to point out the Secret being referenced must be placed in the same Namespace where the SBOMbastic stack is deployed

Author: agree

The user can use these two formats for secrets, since VEX Hub supports auth
through `username`/`password` or `token`.

Below you will find an example of `Secret` with `username` and `password`:

```yaml
# Example Secret for Basic Authentication
apiVersion: v1
kind: Secret
metadata:
  name: company-vexhub-creds
  namespace: default
type: Opaque
data:
  username: YWRtaW4=
  password: c2VjcmV0cGFzc3dvcmQ=
```

Here's how the `Secret` with `token` should look like:

```yaml
# Example Secret for Bearer Token
apiVersion: v1
kind: Secret
metadata:
  name: company-vexhub-token
  namespace: default
type: Opaque
data:
  token: ZXlKaGaU9pSktWMVFpT2c9PQ==
```

To ensure the validity of the input, we are going to provide a 
`ValidatingWebhook` to ensure that the referenced `Secret` has the right form.

## AirGap

AirGap is available by default for this feature, since the only requirement is

> to provide a self-hosted vexhub repository and change the `repository_url` 

Reviewer 2: ```suggestion
to provide a self-hosted VEX Hub repository and change the `repository_url`
```

(if any) within the VEX files, to point to the internal registries.

> This setup is well described here: https://github.com/aquasecurity/trivy/blob/main/docs/docs/advanced/air-gap.md#vex-hub

Reviewer 2: ```suggestion
This setup is well described [here](https://github.com/aquasecurity/trivy/blob/main/docs/docs/advanced/air-gap.md#vex-hub).
```

resources we want to use to scan it.

Suppose you are scanning the registry of a dedicated department in your company.
This registry will host specific kind of images depending on the scope of the 
department. In order to make the scan as much accurate as possible, we are 
going to provide a new clusterwide CRD called VexHub. This CRD will hold 
configuration of the VexHub repositories, so that the Registry CRD will refer 
to these VexHub resources.

Since VexHub CRD is clusterwide, this means that you can use the same 
configuration across multiple registries.

## VexHub CRD

To configure a new VexHub, the user will need to apply a new manifest as follow:

```yaml
apiVersion: sbombastic.rancher.io/v1alpha1
kind: VexHub
metadata:
  name: vendor-vexhub
  namespace: default
spec:
  url: https://vexhub.vendor.com
  credentials:
    secretRef:
      name: vendor-vexhub-secret
      key: access-token
```

As you can see the VexHub CRD has a `secretRef` field, which means that you 
need to provide a Secret to configure the credentials (if any):

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: vexhub-auth
  namespace: security-team
type: Opaque
data:

>   token: <base64-encoded-token>

Reviewer: or username and password: https://trivy.dev/latest/docs/supply-chain/vex/repo/#authentication

> ```

Reviewer: An user might want to provide its own certificate to verify https connections to the VexHUB, we need to research if we can use `SSL_CERT_DIR`. See:  https://github.com/aquasecurity/trivy/discussions/4194#discussioncomment-5828069 and https://pkg.go.dev/crypto/x509#SystemCertPool)

> ## Registry CRD

Reviewer: We agreed that VexHUBs are used globally, so there is no need to configure/ changethe Registry CRD. remove this section

To configure the Registry to use the VexHub repository, the user will need to 
update the Registry manifest as follow:

```yaml
apiVersion: scanner.rancher.io/v1alpha1
kind: Registry
metadata:
  name: registry-example
  namespace: default
spec:
  url: "https://registry-1.docker.io"
  type: "docker"
  auth:
    secretName: "registry-secret"
  discoveryPeriod: "1h"
  scanPeriod: "1d"
  repositories:
    - "repo1"
    - "repo2"
  vexHubRefs:
    - name: vendor-vexhub
    - name: internal-vexhub # example of multiple VexHub configuration
```

This way the registry will be scanned with appropriate VEX files.

## AirGap

> AirGap is available by default for this feature, since the only requirement is

Reviewer: add link to the airgap vexhub guide https://github.com/aquasecurity/trivy/blob/main/docs/docs/advanced/air-gap.md#vex-hub

to provide a self-hosted vexhub repository and change the `repository_url` 
(if any) whitin the VEX files, to point to the internal registries.

# Drawbacks

[drawbacks]: #drawbacks

<!---
Why should we **not** do this?

  * obscure corner cases
  * will it impact performance?
  * what other parts of the product will be affected?
  * will the solution be hard to maintain in the future?
--->

# Alternatives

[alternatives]: #alternatives

<!---
- What other designs/options have been considered?
- What is the impact of not doing this?
--->

> The alternative approach is to allow the VexHub configuration directly into the

Reviewer: as pointed out above, no need to do this
